fn/list.go

```diff
@@ -0,0 +1,279 @@
+// Copyright (c) 2009 The Go Authors. All rights reserved.
+// Copyright (c) 2024 Lightning Labs and the Lightning Network Developers
+
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+
+//    * Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//    * Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following disclaimer
+// in the documentation and/or other materials provided with the
+// distribution.
+//    * Neither the name of Google Inc. nor the names of its
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+package fn
+
+type Node[A any] struct {
+	// prev is a pointer to the previous node in the List.
+	prev *Node[A]
+
+	// next is a pointer to the next node in the List.
+	next *Node[A]
+
+	// list is the root pointer to the List in which this node is located.
+	list *List[A]
+
+	// Value is the actual data contained within the Node.
+	Value A
+}
+
+// Next returns the next list node or nil.
+func (e *Node[A]) Next() *Node[A] {
+	if e.next == &e.list.root {
+		return nil
+	}
+
+	return e.next
+}
+
+// Prev returns the previous list node or nil.
+func (e *Node[A]) Prev() *Node[A] {
+	if e.prev == &e.list.root {
+		return nil
+	}
+
+	return e.prev
+}
+
+// List represents a doubly linked list.
+// The zero value for List is an empty list ready to use.
+type List[A any] struct {
```

Comment:
hmm there's already `lru.List` and `lru.Cache` tho, and we would prefer `lru.Cache` as it's memory safe.

Reply:
I couldn't find "List". I did find [`lru.Cache[K, V any]`](https://pkg.go.dev/github.com/hashicorp/golang-lru/v2#Cache), but I fail to understand precisely how that substitutes here.

Reply:
Yeah I ripped this in the LRU package for neutrino: https://github.com/lightninglabs/neutrino/blob/master/cache/lru/list.go